### PR TITLE
feat: Karpenter provisioners

### PR DESCRIPTION
**Pull request by andrewhibbert**
### What does this PR do?

This PR enables management of Karpenter provisioners via terraform to be used alongside the karpenter addon module.

### Motivation

Wanted to manage Karpenter provisioners from code and be able to use terraform features (interpolation, data lookups etc) to do this.

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [x] Yes, I have added a new example under [examples](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR for add-ons under [add-ons](https://github.com/aws-samples/eks-blueprints-add-ons) repo (if applicable)
- [ ] Yes, I have updated the [docs](https://github.com/aws-ia/terraform-aws-eks-blueprints/tree/main/docs) for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR

**Note**: Not all the PRs required examples and docs except a new pattern or add-on added.

### For Moderators

- [ ] E2E Test successfully complete before merge?

### Additional Notes

* Ran the examples/karpenter code
* Created the deployments in sample_deployments folder
* Pods are running
```
$ kubectl get pods -o wide
NAME                                              READY   STATUS    RESTARTS   AGE     IP            NODE                                        NOMINATED NODE   READINESS GATES
ecsdemo-nodejs-79844c6b8c-hjcsv                   1/1     Running   0          11m     10.0.11.103   ip-10-0-11-191.us-west-2.compute.internal   <none>           <none>
ecsdemo-nodejs-79844c6b8c-qpcrz                   1/1     Running   0          11m     10.0.11.109   ip-10-0-11-191.us-west-2.compute.internal   <none>           <none>
ecsdemo-nodejs-79844c6b8c-wznjt                   1/1     Running   0          11m     10.0.11.240   ip-10-0-11-191.us-west-2.compute.internal   <none>           <none>
ecsdemo-nodejs-custom-userdata-548fb88f6c-98crl   1/1     Running   0          11m     10.0.11.148   ip-10-0-11-123.us-west-2.compute.internal   <none>           <none>
ecsdemo-nodejs-custom-userdata-548fb88f6c-9cfv6   1/1     Running   0          11m     10.0.11.158   ip-10-0-11-123.us-west-2.compute.internal   <none>           <none>
ecsdemo-nodejs-custom-userdata-548fb88f6c-tsmv7   1/1     Running   0          11m     10.0.11.98    ip-10-0-11-123.us-west-2.compute.internal   <none>           <none>
ecsdemo-nodejs-lt-58c579c477-p6wd4                1/1     Running   0          2m49s   10.0.11.96    ip-10-0-11-72.us-west-2.compute.internal    <none>           <none>
ecsdemo-nodejs-lt-58c579c477-q9ksc                1/1     Running   0          2m48s   10.0.11.232   ip-10-0-11-72.us-west-2.compute.internal    <none>           <none>
```
* The nodes:
```
$ kubectl get nodes --show-labels
NAME                                        STATUS   ROLES    AGE     VERSION                LABELS
ip-10-0-11-123.us-west-2.compute.internal   Ready    <none>   12m     v1.22.10-eks-7dc61e8   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=c5a.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-west-2,failure-domain.beta.kubernetes.io/zone=us-west-2b,k8s.io/cloud-provider-aws=8ef806446c9a5b779aee2d589b3df676,karpenter.k8s.aws/instance-category=c,karpenter.k8s.aws/instance-cpu=4,karpenter.k8s.aws/instance-family=c5a,karpenter.k8s.aws/instance-generation=5,karpenter.k8s.aws/instance-hypervisor=nitro,karpenter.k8s.aws/instance-memory=8192,karpenter.k8s.aws/instance-pods=58,karpenter.k8s.aws/instance-size=xlarge,karpenter.sh/capacity-type=spot,karpenter.sh/initialized=true,karpenter.sh/provisioner-name=custom-userdata,kubernetes.io/arch=amd64,kubernetes.io/hostname=ip-10-0-11-123.us-west-2.compute.internal,kubernetes.io/os=linux,node.kubernetes.io/instance-type=c5a.xlarge,provisioner=custom-userdata,topology.kubernetes.io/region=us-west-2,topology.kubernetes.io/zone=us-west-2b,type=karpenter
ip-10-0-11-191.us-west-2.compute.internal   Ready    <none>   12m     v1.22.12-eks-ba74326   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=c5a.xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-west-2,failure-domain.beta.kubernetes.io/zone=us-west-2b,k8s.io/cloud-provider-aws=8ef806446c9a5b779aee2d589b3df676,karpenter.k8s.aws/instance-category=c,karpenter.k8s.aws/instance-cpu=4,karpenter.k8s.aws/instance-family=c5a,karpenter.k8s.aws/instance-generation=5,karpenter.k8s.aws/instance-hypervisor=nitro,karpenter.k8s.aws/instance-memory=8192,karpenter.k8s.aws/instance-pods=58,karpenter.k8s.aws/instance-size=xlarge,karpenter.sh/capacity-type=spot,karpenter.sh/initialized=true,karpenter.sh/provisioner-name=default,kubernetes.io/arch=amd64,kubernetes.io/hostname=ip-10-0-11-191.us-west-2.compute.internal,kubernetes.io/os=linux,node.kubernetes.io/instance-type=c5a.xlarge,provisioner=default,topology.kubernetes.io/region=us-west-2,topology.kubernetes.io/zone=us-west-2b,type=karpenter
ip-10-0-11-72.us-west-2.compute.internal    Ready    <none>   4m21s   v1.22.12-eks-ba74326   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m5.2xlarge,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-west-2,failure-domain.beta.kubernetes.io/zone=us-west-2b,k8s.io/cloud-provider-aws=8ef806446c9a5b779aee2d589b3df676,karpenter.k8s.aws/instance-category=m,karpenter.k8s.aws/instance-cpu=8,karpenter.k8s.aws/instance-family=m5,karpenter.k8s.aws/instance-generation=5,karpenter.k8s.aws/instance-hypervisor=nitro,karpenter.k8s.aws/instance-memory=32768,karpenter.k8s.aws/instance-pods=58,karpenter.k8s.aws/instance-size=2xlarge,karpenter.sh/capacity-type=spot,karpenter.sh/initialized=true,karpenter.sh/provisioner-name=default-lt,kubernetes.io/arch=amd64,kubernetes.io/hostname=ip-10-0-11-72.us-west-2.compute.internal,kubernetes.io/os=linux,node.kubernetes.io/instance-type=m5.2xlarge,provisioner=default-lt,topology.kubernetes.io/region=us-west-2,topology.kubernetes.io/zone=us-west-2b,type=karpenter
ip-10-0-12-79.us-west-2.compute.internal    Ready    <none>   117m    v1.22.12-eks-ba74326   beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=m5.large,beta.kubernetes.io/os=linux,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/nodegroup-image=ami-09f0868e933b1494d,eks.amazonaws.com/nodegroup=managed-ondemand-2022081718304867890000000e,eks.amazonaws.com/sourceLaunchTemplateId=lt-079277b863016d6ae,eks.amazonaws.com/sourceLaunchTemplateVersion=1,failure-domain.beta.kubernetes.io/region=us-west-2,failure-domain.beta.kubernetes.io/zone=us-west-2c,k8s.io/cloud-provider-aws=8ef806446c9a5b779aee2d589b3df676,kubernetes.io/arch=amd64,kubernetes.io/hostname=ip-10-0-12-79.us-west-2.compute.internal,kubernetes.io/os=linux,node.kubernetes.io/instance-type=m5.large,topology.kubernetes.io/region=us-west-2,topology.kubernetes.io/zone=us-west-2c
```
* Custom user data instance is a bottlerocket one
customuserdatainstance

### CODE REVIEW

File: modules/aws-eks-karpenter-provisioners/variables.tf [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    key    = string
+    effect = string
+  }))
+  default = []
+}
+
+variable "ttl_seconds_after_empty" {
+  description = "How long after a node is empty will it be scaled down due to low utilization. If null, the feature is disabled."
+  type        = number
+  default     = 30
```
Comment (FernandoMiguel) [inline modules/aws-eks-karpenter-provisioners/variables.tf:141]: this is too low, given how long a new one takes to spin up, and pricing being 60 seconds initially.

**Comment by bryantbiggs** [comment]
Hi @andrewhibbert - can you please take a look at the contributing guidelines when you get a chance

File: modules/aws-eks-karpenter-provisioners/variables.tf (hunk 1 above)
Comment (andrewhibbert) [reply modules/aws-eks-karpenter-provisioners/variables.tf]: I have made this `null`

**Comment by andrewhibbert** [comment]
> Hi @andrewhibbert - can you please take a look at the contributing guidelines when you get a chance

I have made some changes to `examples/karpenter` an updated the checklist

**Comment by bryantbiggs** [comment]
> > Hi @andrewhibbert - can you please take a look at the contributing guidelines when you get a chance
> 
> I have made some changes to `examples/karpenter` and updated the checklist

Its more around opening an issue first for substantial changes or changes in direction. This change to create a module for Karpenter provisioners is interesting but we should discuss first. The motivation for such a change, the problem(s) or challenge(s) it solves, how it fits into the bigger picture of what Blueprints provides, etc.

**Comment by bryantbiggs** [comment]
thank you for opening this - however, we won't be proceeding with support for this approach. users can continue to use the YAML manifests for creating provisioners. thank you

**Pull request closed without merging**